### PR TITLE
Add TS types and tests

### LINKED ISSUE #38: Add TypeScript definitions
## 🚀 Feature Proposal

Add an index.d.ts with Typescript type definitions.

## Motivation

To use it on Typescript projects.

I can do a PR for this if you are ok with it.

### PR DESCRIPTION
**Pull request by ricardo-devis-agullo**
Closes  #38 
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+declare module "fastify" {
+  interface FastifyInstance<HttpServer, HttpRequest, HttpResponse> {
+    redis: ioredis.Redis;
+  }
+}
+
+declare const fastifyBasicAuth: fastify.Plugin<
+  Server,
+  IncomingMessage,
+  ServerResponse,
```
Comment (mcollina) [inline index.d.ts]: Would you mind using https://github.com/fastify/fastify-swagger/blob/2f7d7f68f8d87f16904b5da8bec38784b64ac5c8/index.d.ts#L70-L73 so we support http2 as well?

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+// Definitions by: ricardo-devis-agullo <https://github.com/ricardo-devis-agullo>
```
Comment (mcollina) [inline index.d.ts]: Would you mind removing this? The attribution will stay forever in the git commit log, as every other change in this repo.

**Comment by ricardo-devis-agullo** [comment]
package-lock removed!

File: index.d.ts (hunk 1 above)
Comment (ricardo-devis-agullo) [reply index.d.ts]: hmm this one seems it breaks the tests as TS can´t find a viable type that its a union of both http and http2 servers...

Comment (mcollina) [reply index.d.ts]: Check out fastify-swagger, it supports both... so it should be doable.

Comment (ricardo-devis-agullo) [reply index.d.ts]: I just checked and I'm not sure they are correctly set. I just changed something on the tests to an invalid type, and the typescript compiler passed!

Typechecker not working

**Comment by mcollina** [comment]
CI is failing.

**Comment by nboltn** [comment]
why is this still open?

**Comment by mcollina** [comment]
@nboltn because CI is failing. Would you like to send a fresh PR?

**Pull request closed without merging**